ocaml/xapi/xapi_pusb_helpers.ml

```diff
@@ -0,0 +1,112 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+module D=Debug.Make(struct let name="xapi" end)
+open D
+
+open Stdext
+open Yojson.Basic.Util
+
+module USB = struct
+  type t = {
+    path         : string;
+    vendor_id    : string;
+    vendor_desc  : string;
+    product_id   : string;
+    product_desc : string;
+    serial       : string;
+    version      : string;
+    description  : string;
+  }
+  let compare x y =
+    if x.path = y.path then
+      if x.vendor_id = y.vendor_id then
+        if x.product_id = y.product_id then
+          compare x.serial y.serial
+        else
+          compare x.product_id y.product_id
+      else
+        compare x.vendor_id y.vendor_id
+    else
+      compare x.path y.path
+end
+
+module USBSet = Set.Make(USB)
+
+let extract_member json member =
+  [json]
+    |> filter_member member
+    |> filter_string
+
+let extract_local_usb_info usb =
+  let open USB in
+  {
+    path = List.hd (extract_member usb "path");
```

Comment:
These unguarded `List.hd` calls are a bit worrying - they've been a horrible source of hard-to-trace exceptions in the past. I suggest you replace `extract_member` with something more like:
```ocaml
let extract_member json member =
    let safe_hd = function
    | x::_ -> x
    | _ -> failwith (Printf.sprintf "Internal error: Json from scan script missing element: %s" member)
    in
    [json] |> filter_member member |> filter_string |> safe_hd
```

Reply:
Ok, make sense.

Reply:
updated @jonludlam